waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/TunnelingMetaStoreClientFactoryTest.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (C) 2016-2017 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.waggledance.client;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.when;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import com.pastdev.jsch.tunnel.TunnelConnectionManager;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({ TunnelHandler.class, TunnelingMetaStoreClientFactory.class })
+public class TunnelingMetaStoreClientFactoryTest {
+
+  private @Spy TunnelingMetaStoreClientFactory tunnelingMetaStoreClientFactory = new TunnelingMetaStoreClientFactory();
+  private @Mock CloseableThriftHiveMetastoreIface client;
+  private @Mock TunnelConnectionManager tunnelConnectionManager;
+  private @Spy HiveConf hiveConf = new HiveConf();
+  private @Spy HiveConf localHiveConf = new HiveConf();
+
+  @Test
+  public void newInstanceWithTunneling() throws Exception {
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_ROUTE.varname, "user@ec2-12-345-678-91.compute-1.amazonaws.com");
+    hiveConf.set(WaggleDanceHiveConfVars.SSH_PRIVATE_KEYS.varname, "private_key");
+    hiveConf.setVar(HiveConf.ConfVars.METASTOREURIS,
+        "thrift://internal-test-shared-hive-metastore-elb-1234567891.us-west-1.elb.amazonaws.com:1234");
+
+    PowerMockito.mockStatic(TunnelHandler.class);
```

Comment:
TunnelHandler should be a constructor argument that is mocked in this test. Then you can verify interactions with the tunnelhandler better

Reply:
Lol, I did this originally then was told to change it to a static method, I'm going to leave it for now, anyone else can chip in their thoughts until their is consensus (all the small changes are slowing down this review)

Reply:
I would not use PowerMock unless it is the only way to test something. e.g. legacy code, third party stuff. Leads to very tightly coupled code and brittle tests. I understand this tunneling stuff is hard to test, just saying use with care. But sure a bad test is still better than no test :).

This is your PR, you should take our comments into account and decide for yourself. This is not a massive thing and I don't want to go back and forth so I'll approve and leave it to you.

BTW: I tried to find the comment you refer to or the code change but I can't really find it so not sure what the original code was that triggered you to change it to this, it all seems to have changed quite a bit since then anyway.

Reply:
I don't know why it's been changed but you the problem here is the class that connects to the SSH. If you manage to provide a mock instance of `TunnelConnectionManager` this static method is not needed

Reply:
nah mock TunnelHandler I meant

Reply:
How can you inject TunnelConnectionManagerFactory? The interface doesn't allow it to be injected and the constructor for TunnelConnectionManagerFactory requires the relevant HiveConf passed into newInstance? TunnelHandler can become a concrete class, but we will still have to use power mock to write a test that even captures the bug that this PR fixes, so if thats the only reason to make it not static then I don't really see the point. Power mock was a last resort to work around how this class is tied to the MetaStoreClientFactory interface, I can see why this class didn't have tests originally after working with it.